File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

// Helper structure to mimic returning a 2-element array [sum, count]
struct SubtreeData {
    int sum;
    int count;
};

struct SubtreeData dfs(struct TreeNode* node, int* answer) {
    if (node == NULL) {
        struct SubtreeData base = {0, 0};
        return base;
    }

    struct SubtreeData left = dfs(node->left, answer);
    struct SubtreeData right = dfs(node->right, answer);

    struct SubtreeData current;
    current.sum = node->val + left.sum + right.sum;
    current.count = 1 + left.count + right.count;

    if (current.sum / current.count == node->val) {
        (*answer)++;
    }

    return current;
}

int averageOfSubtree(struct TreeNode* root) {
    int answer = 0;
    dfs(root, &answer);
    return answer;
}

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var averageOfSubtree = function(root) {
    let answer = 0;

    function dfs(node) {
        // FIXED: Now correctly returns the base array structure
        if (node === null) {
            return [0, 0]; 
        }

        const left = dfs(node.left);
        const right = dfs(node.right);

        const sum = node.val + left[0] + right[0];
        const count = 1 + left[1] + right[1];

        if (Math.floor(sum / count) === node.val) {
            answer++;
        }

        return [sum, count];
    }

    dfs(root);
    return answer;
};

